lib/cPanel/StateFile.pm

```diff
@@ -253,26 +269,27 @@ sub import {
             $state_file->throw('Cannot update_file inside a call_unlocked call.') unless defined $self->{lock_file};
 
             if ( !$state_file->{file_handle} ) {
-                if ( -e $state_file->{file_name} ) {
-                    $self->_open('+<');
-                }
-                else {
-                    sysopen( my $fh, $state_file->{file_name}, &Fcntl::O_CREAT | &Fcntl::O_EXCL | &Fcntl::O_RDWR )
-                      or $state_file->throw("Cannot create state file '$state_file->{file_name}': $!");
-                    $state_file->{file_handle} = $fh;
-                }
+                $self->_open('+<');
             }
-            seek( $state_file->{file_handle}, 0, 0 );
-            truncate( $state_file->{file_handle}, 0 )
-              or $state_file->throw("Unable to truncate the state file: $!");
 
-            $state_file->{data_object}->save_to_cache( $state_file->{file_handle} );
-            $state_file->{file_mtime} = ( stat( $state_file->{file_handle} ) )[9];
+            #Set UNLINK in case we die().
+            require File::Temp;
+            my ($fh, $path) = File::Temp::tempfile( DIR => $self->{_file_dir}, UNLINK => 1 );
+
+            $self->_flock_after_open($fh);
```

Comment:
Here you're locking the temporary file, which doesn't seem to prevent concurrent writes by multiple processes, since each temporary file will by definition have a different name.

Reply:
The lock on the temp file is so that when the temp file is moved into place it’s “pre-locked”. I’ll comment about that.